<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;">Loading libraries

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;">Exploratory data analysis

In [ ]:
data = pd.read_csv('bank_marketing_part1_Data.csv')
data.head()

spending  advance_payments  probability_of_full_payment  current_balance  \
0     19.94             16.92                       0.8752            6.675   
1     15.99             14.89                       0.9064            5.363   
2     18.95             16.42                       0.8829            6.248   
3     10.83             12.96                       0.8099            5.278   
4     17.99             15.86                       0.8992            5.890   

   credit_limit  min_payment_amt  max_spent_in_single_shopping  
0         3.763            3.252                         6.550  
1         3.582            3.336                         5.144  
2         3.755            3.368                         6.148  
3         2.641            5.182                         5.185  
4         3.694            2.068                         5.837

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;"> Checking the Dimension of data

In [ ]:
data.shape

(210, 7)

**There are 210 Observations and 7 Attributes.**

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;">Checking the information about data

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 210 entries, 0 to 209
Data columns (total 7 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   spending                      210 non-null    float64
 1   advance_payments              210 non-null    float64
 2   probability_of_full_payment   210 non-null    float64
 3   current_balance               210 non-null    float64
 4   credit_limit                  210 non-null    float64
 5   min_payment_amt               210 non-null    float64
 6   max_spent_in_single_shopping  210 non-null    float64
dtypes: float64(7)
memory usage: 11.6 KB


<span style="font-family: Arial; font-weight:bold;font-size:1.7em;color:#00b3e5;">EDA Descriptive Statistics

### **Mean**

In [ ]:
data.mean()

spending                        14.847524
advance_payments                14.559286
probability_of_full_payment      0.870999
current_balance                  5.628533
credit_limit                     3.258605
min_payment_amt                  3.700201
max_spent_in_single_shopping     5.408071
dtype: float64

### **Median**

In [ ]:
data.median()

spending                        14.35500
advance_payments                14.32000
probability_of_full_payment      0.87345
current_balance                  5.52350
credit_limit                     3.23700
min_payment_amt                  3.59900
max_spent_in_single_shopping     5.22300
dtype: float64

### **Mode**

In [ ]:
data.mode()

spending  advance_payments  probability_of_full_payment  current_balance  \
0     11.23             13.47                       0.8823            5.236   
1     14.11               NaN                          NaN            5.395   
2     15.38               NaN                          NaN              NaN   

   credit_limit  min_payment_amt  max_spent_in_single_shopping  
0         3.026            2.129                         5.001  
1           NaN            2.221                           NaN  
2           NaN            2.700                           NaN

### Quantile 25%

In [ ]:
print('data_quantile(25%):',data.quantile(q=0.25))

data_quantile(25%): spending                        12.27000
advance_payments                13.45000
probability_of_full_payment      0.85690
current_balance                  5.26225
credit_limit                     2.94400
min_payment_amt                  2.56150
max_spent_in_single_shopping     5.04500
Name: 0.25, dtype: float64


### Quantile 50%

In [ ]:
print('data_quantile(50%):',data.quantile(q=0.50))

data_quantile(50%): spending                        14.35500
advance_payments                14.32000
probability_of_full_payment      0.87345
current_balance                  5.52350
credit_limit                     3.23700
min_payment_amt                  3.59900
max_spent_in_single_shopping     5.22300
Name: 0.5, dtype: float64


### Quantile 75%

In [ ]:
print('data_quantile(75%):',data.quantile(q=0.75))

NameError: ignored

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;">Checking the Measures of Dispersion

In [ ]:
# IQR

data.quantile(0.75)-data.quantile(0.25)

In [ ]:
#  Range

data.max()-data.min()

In [ ]:
#  Variance

data.var()

In [ ]:
# standard deviation

data.std()

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;"> Plotting the summary mean,mode,median using histogram

In [ ]:
mean = data['spending'].mean()
median = data['spending'].median()
mode = data['spending'].mode()
print('Mean: ',mean,'\nMedian: ',median,'\nMode: ',mode[0])

plt.figure(figsize=(10,5))
plt.hist(data['spending'],bins=100,color='lightblue') #plot histogram

# draw lines for mean,mode,median

plt.axvline(mean,color='green',label='mean')
plt.axvline(median,color='blue',label='median')
plt.axvline(mode[0],color='red',label='mode 1')
plt.axvline(mode[1],color='red',label='mode 2')
plt.axvline(mode[2],color='red',label='mode 3')

# label x axis

plt.xlabel('spending')

# label y axis

plt.ylabel('Frequency')

# plot legend

plt.legend()

plt.show()

In [ ]:
mean = data['advance_payments'].mean()
median = data['advance_payments'].median()
mode = data['advance_payments'].mode()
print('Mean: ',mean,'\nMedian: ',median,'\nMode: ',mode[0])

plt.figure(figsize=(10,5))
plt.hist(data['advance_payments'],bins=100,color='lightblue') #plot histogram

# draw lines for mean,mode,median

plt.axvline(mean,color='green',label='mean')
plt.axvline(median,color='blue',label='median')
plt.axvline(mode[0],color='red',label='mode')

# label x axis

plt.xlabel('advance_payments')

# label y axis

plt.ylabel('Frequency')

# plot legend

plt.legend()

plt.show()

In [ ]:
mean = data['probability_of_full_payment'].mean()
median = data['probability_of_full_payment'].median()
mode = data['probability_of_full_payment'].mode()
print('Mean: ',mean,'\nMedian: ',median,'\nMode: ',mode[0])

plt.figure(figsize=(10,5))
plt.hist(data['probability_of_full_payment'],bins=100,color='lightblue') #plot histogram

# draw lines for mean,mode,median

plt.axvline(mean,color='green',label='mean')
plt.axvline(median,color='blue',label='median')
plt.axvline(mode[0],color='red',label='mode')

# label x axis

plt.xlabel('probability_of_full_payment')

# label y axis

plt.ylabel('Frequency')

# plot legend

plt.legend()

plt.show()

In [ ]:
mean = data['current_balance'].mean()
median = data['current_balance'].median()
mode = data['current_balance'].mode()
print('Mean: ',mean,'\nMedian: ',median,'\nMode: ',mode[0])

plt.figure(figsize=(10,5))
plt.hist(data['current_balance'],bins=100,color='lightblue') #plot histogram

# draw lines for mean,mode,median

plt.axvline(mean,color='green',label='mean')
plt.axvline(median,color='blue',label='median')
plt.axvline(mode[0],color='red',label='mode 1')
plt.axvline(mode[1],color='red',label='mode 2')


# label x axis

plt.xlabel('current_balance')

# label y axis

plt.ylabel('Frequency')

# plot legend

plt.legend()

plt.show()

In [ ]:
mean = data['credit_limit'].mean()
median = data['credit_limit'].median()
mode = data['credit_limit'].mode()
print('Mean: ',mean,'\nMedian: ',median,'\nMode: ',mode[0])

plt.figure(figsize=(10,5))
plt.hist(data['credit_limit'],bins=100,color='lightblue') #plot histogram

# draw lines for mean,mode,median

plt.axvline(mean,color='green',label='mean')
plt.axvline(median,color='blue',label='median')
plt.axvline(mode[0],color='red',label='mode 1')


# label x axis

plt.xlabel('credit_limit')

# label y axis

plt.ylabel('Frequency')

# plot legend

plt.legend()

plt.show()

In [ ]:
mean = data['min_payment_amt'].mean()
median = data['min_payment_amt'].median()
mode = data['min_payment_amt'].mode()
print('Mean: ',mean,'\nMedian: ',median,'\nMode: ',mode[0])

plt.figure(figsize=(10,5))
plt.hist(data['min_payment_amt'],bins=100,color='lightblue') #plot histogram

# draw lines for mean,mode,median

plt.axvline(mean,color='green',label='mean')
plt.axvline(median,color='blue',label='median')
plt.axvline(mode[0],color='red',label='mode 1')
plt.axvline(mode[1],color='red',label='mode 2')
plt.axvline(mode[2],color='red',label='mode 3')

# label x axis

plt.xlabel('min_payment_amt')

# label y axis

plt.ylabel('Frequency')

# plot legend

plt.legend()

plt.show()

In [ ]:
mean = data['max_spent_in_single_shopping'].mean()
median = data['max_spent_in_single_shopping'].median()
mode = data['max_spent_in_single_shopping'].mode()
print('Mean: ',mean,'\nMedian: ',median,'\nMode: ',mode[0])

plt.figure(figsize=(10,5))
plt.hist(data['max_spent_in_single_shopping'],bins=100,color='lightblue') #plot histogram

# draw lines for mean,mode,median

plt.axvline(mean,color='green',label='mean')
plt.axvline(median,color='blue',label='median')
plt.axvline(mode[0],color='red',label='mode 1')

# label x axis

plt.xlabel('max_spent_in_single_shopping')

# label y axis

plt.ylabel('Frequency')

# plot legend

plt.legend()

plt.show()

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;">Checking Covariance and Correlation

In [ ]:
data.cov()

In [ ]:
data.corr()

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;">Pairplot for checking the correlation

In [ ]:
import seaborn as sns
sns.pairplot(data,kind='reg'); #plots scatter plots for every pair of attributes and histograms along the diagonal
plt.show()

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;"> Heatmap For checking the correlation

In [ ]:
fig,ax = plt.subplots(figsize=(10, 10))
sns.heatmap(data.corr(), ax=ax, annot=True, linewidths=0.05, fmt= '.2f',cmap="magma")
plt.show()

There is some negative correlation between variables. spending and advance_payments are highly correlated. spending and credit_limit are also highly correlated.advance_payments and current_balance are also strongly correlated.probability_of_payment and min_payment_amt are negatively and least correlated.

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;">Checking the Skewness

In [ ]:
data.skew()

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;"> Checking the missing values

In [ ]:
data.isnull().sum()

**no missing values**

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;"> Checking the duplicate values

In [ ]:
data.duplicated().sum()

**No duplicate values**

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;"> Checking for outliers

In [ ]:
# construct box plot for continuous variables
plt.figure(figsize=(15,15))
data[['spending', 'advance_payments',
    'probability_of_full_payment', 'current_balance', 'credit_limit',
      'min_payment_amt', 'max_spent_in_single_shopping']].boxplot(vert=0)

**There are outliers in min_payment_amt and probability_of_full_payment**

In [ ]:
data_copied = data.copy()
data_copied.head()

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;"> 1.2  Do you think scaling is necessary for clustering in this case? Justify
    

In [ ]:
data.describe()

**By checking the summary, we can say that min,max and mean are not equal for all the columns.scaling is necessary**

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;"> scaling the data

In [ ]:
from sklearn.preprocessing import MinMaxScaler

minmax_scale = MinMaxScaler()

minmax_scale

In [ ]:
data_copied['spending_MinMaxScale'] = minmax_scale.fit_transform(data_copied[['spending']])
data_copied['spending_MinMaxScale'].head(5)

In [ ]:
data_copied['advance_payments_MinMaxScale'] = minmax_scale.fit_transform(data_copied[['advance_payments']])
data_copied['advance_payments_MinMaxScale'].head(5)

In [ ]:
data_copied['probability_of_full_payment_MinMaxScale'] = minmax_scale.fit_transform(data_copied[['probability_of_full_payment']])
data_copied['probability_of_full_payment_MinMaxScale'].head(5)

In [ ]:
data_copied['current_balance_MinMaxScale'] = minmax_scale.fit_transform(data_copied[['current_balance']])
data_copied['current_balance_MinMaxScale'].head(5)

In [ ]:
data_copied['credit_limit_MinMaxScale'] = minmax_scale.fit_transform(data_copied[['credit_limit']])
data_copied['credit_limit_MinMaxScale'].head(5)

In [ ]:
data_copied['min_payment_amt_MinMaxScale'] = minmax_scale.fit_transform(data_copied[['min_payment_amt']])
data_copied['min_payment_amt_MinMaxScale'].head(5)

In [ ]:
data_copied['max_spent_in_single_shopping_MinMaxScale'] = minmax_scale.fit_transform(data_copied[['max_spent_in_single_shopping']])
data_copied['max_spent_in_single_shopping_MinMaxScale'].head(5)

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;">Applying hierarchical clustering to scaled data

In [ ]:
from scipy.cluster.hierarchy import dendrogram, linkage

In [ ]:
wardlink = linkage(data_copied, method = 'ward')

In [ ]:
dendrogram(wardlink)

In [ ]:
dend = dendrogram(wardlink)

In [ ]:
dend = dendrogram(wardlink, truncate_mode = 'lastp', p =10)

**p value shows the last 10 merges. Green cluster has maximum products, the orange cluster has minimum products**

**we created the linkage, that is the distances. Now obtaining the clusters and products belong to each cluster for final verification**

In [ ]:
from scipy.cluster.hierarchy import fcluster

we will use these fcluster to form the clusters

In [ ]:
clusters = fcluster(wardlink, 23, criterion = 'distance')
clusters

In [ ]:
data_copied['clusters'] = clusters

In [ ]:
data_copied.head()

In [ ]:
data_copied.to_csv('hc.csv')

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;">Applying K-Means clustering on scaled data


In [ ]:
data.head()

In [ ]:
data.shape

<span style="font-family: Arial; font-weight:bold;font-size:1.5em;color:#00b3e5;">scaling the data

In [ ]:
from sklearn.preprocessing import StandardScaler

In [ ]:
X = StandardScaler()

In [ ]:
scaled_df = X.fit_transform(data)
scaled_df

In [ ]:
# Create K Means cluster and store the result in the object k_means
from sklearn.cluster import KMeans
k_means = KMeans(n_clusters = 2,random_state=1)

In [ ]:
# Creating Clusters using KMeans
### Forming 2 Clusters with K=2

In [ ]:
# Fit K means on the scaled_df
k_means.fit(scaled_df)

In [ ]:
### Cluster Output for all the observations

In [ ]:
# Get the labels
k_means.labels_

In [ ]:
### Within Cluster Sum of Squares

In [ ]:
k_means.inertia_

In [ ]:
### Forming clusters with K = 1,3,4,5,6 and comparing the WSS

In [ ]:
k_means = KMeans(n_clusters = 1,random_state=1)
k_means.fit(scaled_df)
k_means.inertia_

In [ ]:
k_means = KMeans(n_clusters = 3,random_state=1)
k_means.fit(scaled_df)
k_means.inertia_

In [ ]:
k_means = KMeans(n_clusters = 4,random_state=1)
k_means.fit(scaled_df)
k_means.inertia_

In [ ]:
k_means = KMeans(n_clusters = 5,random_state=1)
k_means.fit(scaled_df)
k_means.inertia_

In [ ]:
k_means = KMeans(n_clusters = 6,random_state=1)
k_means.fit(scaled_df)
k_means.inertia_

**WSS reduces as K keeps increasing**

In [ ]:
### Calculating WSS for other values of K - Elbow Method

In [ ]:
wss =[]

In [ ]:
for i in range(1,11):
    KM = KMeans(n_clusters=i,random_state=1)
    KM.fit(scaled_df)
    wss.append(KM.inertia_)

In [ ]:
wss

In [ ]:
a=[1,2,3,4,5,6,7,8,9,10]

In [ ]:
sns.pointplot(a, wss);

In [ ]:
plt.plot(range(1,11), wss)

In [ ]:
# KMeans with K=3

In [ ]:
k_means = KMeans(n_clusters = 3,random_state=1)
k_means.fit(scaled_df)
labels = k_means.labels_

In [ ]:
### Cluster evaluation for 3 clusters: the silhouette score

In [ ]:
from sklearn.metrics import silhouette_samples, silhouette_score

In [ ]:
# Calculating silhouette_score
silhouette_score(scaled_df,labels,random_state=1)

In [ ]:
# KMeans with K=4

In [ ]:
k_means = KMeans(n_clusters = 4,random_state=1)
k_means.fit(scaled_df)
labels = k_means.labels_

In [ ]:
### Cluster evaluation for 4 clusters

In [ ]:
silhouette_score(scaled_df,labels,random_state=1)

**silhouette score is better for 3 clusters than for 4 clusters. So, final clusters will be 3**

Appending Clusters to the original dataset

In [ ]:
data["Clus_kmeans4"] = labels
data.head()

**Cluster profiles for Hierarchial Cluster**

Cluster Frequency

In [ ]:
data_copied.clusters.value_counts().sort_index()

In [ ]:
# Cluster Profiles

In [ ]:
aggdata=data_copied.groupby('clusters').mean()
aggdata['Freq']=data_copied.clusters.value_counts().sort_index()
aggdata

**Cluster profiles for kmeans Cluster**

In [ ]:
data.Clus_kmeans4.value_counts().sort_index()

In [ ]:
clust_profile=data
clust_profile=clust_profile.groupby('Clus_kmeans4').mean()
clust_profile['freq']=data.Clus_kmeans4.value_counts().sort_index()
clust_profile

In [ ]:
data.to_csv('km.csv')